pkg/api/server/v1alpha2/log/gcs.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/tektoncd/results/pkg/apis/v1alpha2"
+
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/gcp"
+)
+
+type gcsStream struct {
+	ctx    context.Context
+	config *server.Config
+	key    string
+	client *gcp.HTTPClient
+}
+
+// NewGCSStream returns a log streamer for the GCS storage type.
+func NewGCSStream(ctx context.Context, log *v1alpha2.Log, config *server.Config) (Stream, error) {
+	if log.Status.Path == "" {
+		filePath, err := FilePath(log)
+		if err != nil {
+			return nil, err
+		}
+		log.Status.Path = filePath
+	}
+
+	filePath := filepath.Join(config.LOGS_PATH, log.Status.Path)
+
+	client, err := getGCSClient(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	gcs := &gcsStream{
+		ctx:    ctx,
+		config: config,
+		key:    filePath,
+		client: client,
+	}
+
+	return gcs, nil
+}
+
+func getGCSClient(ctx context.Context, cfg *server.Config) (*gcp.HTTPClient, error) {
+	var creds *google.Credentials
+
+	if cfg.STORAGE_EMULATOR_HOST != "" {
+		creds, _ = google.CredentialsFromJSON(ctx, []byte(`{"type": "service_account", "project_id": "my-project-id"}`))
+	} else {
+		var err error
+		creds, err = gcp.DefaultCredentials(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return gcp.NewHTTPClient(
+		gcp.DefaultTransport(),
+		gcp.CredentialsTokenSource(creds))
+}
+
+func (*gcsStream) Type() string {
+	return string(v1alpha2.GCSLogType)
+}
+
+func (gcs *gcsStream) WriteTo(w io.Writer) (int64, error) {
+	bucket, err := gcsblob.OpenBucket(gcs.ctx, gcs.client, gcs.config.GCS_BUCKET_NAME, nil)
+	if err != nil {
+		return 0, fmt.Errorf("could not open bucket: %v", err)
+	}
+	defer bucket.Close()
+
+	r, err := bucket.NewReader(gcs.ctx, gcs.key, nil)
+	if err != nil {
+		return 0, fmt.Errorf("could not create bucket reader: %v", err)
+	}
+	n, err := r.WriteTo(w)
+	if err != nil {
+		return 0, fmt.Errorf("could not read from bucket: %v", err)
+	}
+	return n, nil
+}
+
+func (gcs *gcsStream) ReadFrom(r io.Reader) (n int64, err error) {
+	bucket, err := gcsblob.OpenBucket(gcs.ctx, gcs.client, gcs.config.GCS_BUCKET_NAME, nil)
+	if err != nil {
+		return 0, fmt.Errorf("could not open bucket: %v", err)
+	}
+	defer bucket.Close()
+
+	w, err := bucket.NewWriter(gcs.ctx, gcs.key, nil)
+	if err != nil {
+		return 0, fmt.Errorf("could not create bucket writer: %v", err)
+	}
+	defer func() {
+		err = w.Close()
+		if err != nil {
+			err = fmt.Errorf("could not write to bucket(close): %v", err)
+		}
+	}()
+
+	n, err = w.ReadFrom(r)
+	if err != nil {
+		return 0, fmt.Errorf("could not write to bucket: %v", err)
+	}
+	return n, nil
+}
+
+func (gcs *gcsStream) Flush() error {
+	return nil
+}
+
+func (gcs *gcsStream) Delete() error {
+	bucket, err := gcsblob.OpenBucket(gcs.ctx, gcs.client, gcs.config.GCS_BUCKET_NAME, nil)
+	if err != nil {
+		return fmt.Errorf("could not open bucket: %v", err)
+	}
+	defer bucket.Close()
+
+	if err := bucket.Delete(gcs.ctx, gcs.key); err != nil {
+		return fmt.Errorf("could not delete bucket: %v", err)
```

Comment:
Log the key here? (error message implies we couldn't delete the whole bucket!)

Reply:
Done. Fix the error message.